pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/ExpValFunctors.hpp

```diff
@@ -414,14 +416,14 @@ template <class PrecisionT> struct getExpVal3QubitOpFunctor {
     KOKKOS_INLINE_FUNCTION
     void operator()(const std::size_t k, PrecisionT &expval) const {
         const std::size_t kdim = k * dim;
-        INDEX(i000, 0B000);
-        INDEX(i001, 0B001);
-        INDEX(i010, 0B010);
-        INDEX(i011, 0B011);
-        INDEX(i100, 0B100);
-        INDEX(i101, 0B101);
-        INDEX(i110, 0B110);
-        INDEX(i111, 0B111);
+        std::size_t i000 = INDEX(i000, 0B000);
+        std::size_t i001 = INDEX(i001, 0B001);
+        std::size_t i010 = INDEX(i010, 0B010);
+        std::size_t i011 = INDEX(i011, 0B011);
+        std::size_t i100 = INDEX(i100, 0B100);
+        std::size_t i101 = INDEX(i101, 0B101);
+        std::size_t i110 = INDEX(i110, 0B110);
+        std::size_t i111 = INDEX(i111, 0B111);
```

Comment:
I understand that you need to store each one of these index values for further reuse.
I wonder if you could instead of a block of code like that (here and in other places) have a for loop and store the values in an array.

Reply:
Here for example you are essentially going from 0 to 7.

Reply:
> I understand that you need to store each one of these index values for further reuse. I wonder if you could instead of a block of code like that (here and in other places) have a for loop and store the values in an array.

This is similar to the previous (bugged) implementation was doing. The problem is declaring an array, vector or a Kokkos::view is a host function, so it breaks device-backended Kokkos. Now what was done to circumvent that was having a Kokkos::view attribute in the functor. This will compile but since this attribute is shared across threads, they all try writing to it which lead to the bug preventing evaluation of multi-qubit gates. The way to solve this (in the last release) was introducing the three-level team implementation which allows declaring Kokkos::ScratchView's (which are not shared across threads). But this implementation (at least with the default heuristics) is significantly slower than the one proposed here in many cases. 

So long story short, I don't see this working because of Kokkos-related constraints unfortunately.

Reply:
So the only way we could do this is to define a host-side Kokkos view, then pass that in somehow and use that for the updates? Would that be possible, or preferred? If not, at least we know the above generates performant code for now.

Reply:
I assume also these variables are private-level in CUDA, rather than global, so each would be thread-local?

Reply:
Beside `Kokkos::ScratchView`, I don't know how this could work. You could maybe allocate and pass a couple `Kokkos::View`s that have `nthreads*2**nwires` entries and try to read/write only parts of it, but I'm not sure this will work or be more elegant down the line.

Reply:
> I assume also these variables are private-level in CUDA, rather than global, so each would be thread-local?

Yes.